File: chroma-manager/tests/selenium/views/modal.py
from selenium.common.exceptions import NoSuchElementException, StaleElementReferenceException
from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support.expected_conditions import element_to_be_clickable
from selenium.webdriver.common.by import By

from tests.selenium.base_view import BaseView

from tests.selenium.utils.constants import wait_time
from tests.selenium.utils.element import enter_text_for_element
from tests.selenium.utils.fancy_select import FancySelect


class Modal(BaseView):
    @property
    def modal(self):
        return self.driver.find_element_by_css_selector(self.MODAL)

    @property
    def modal_backdrop(self):
        return self.driver.find_element_by_css_selector(self.MODAL_BACKDROP)

    def wait_for_modal(self, wait=wait_time['medium']):
        return WebDriverWait(self, wait).until(lambda modal: modal.modal)

    def wait_for_modal_remove(self):
        def is_removed(obj, prop_name):
            try:
                getattr(obj, prop_name)
            except NoSuchElementException:
                return True

        return WebDriverWait(self, self.medium_wait).until(
            lambda modal: is_removed(modal, 'modal') and is_removed(modal, 'modal_backdrop'),
            "Modal still present after %s seconds." % self.medium_wait)

    def modal_is_displayed(self):
        return self.modal.is_displayed()


class CommandModal(Modal):
    MODAL = '.command-modal'
    MODAL_BACKDROP = '.command-modal-backdrop'
    CLOSE_BUTTON = '%s .btn-danger' % MODAL

    @property
    def close_button(self):
        return self.driver.find_element_by_css_selector(self.CLOSE_BUTTON)

    def wait_for_close_button_to_be_clickable(self):
        return WebDriverWait(self.driver, self.medium_wait).until(
            element_to_be_clickable((By.CSS_SELECTOR, self.CLOSE_BUTTON)), 'Close button not clickable.')


class ConfirmActionModal(Modal):
    MODAL = '.confirm-action-modal'
    MODAL_BACKDROP = '.confirm-action-modal-backdrop'
    CONFIRM_BUTTON = '%s .btn-success' % MODAL

    @property
    def confirm_button(self):
        return self.driver.find_element_by_css_selector(self.CONFIRM_BUTTON)


class AddServerModal(Modal):
    MODAL = '.add-server-modal'
    MODAL_BACKDROP = '.add-server-modal-backdrop'
    MODAL_BODY = '%s .modal-body' % MODAL
    SELECT_SERVER_PROFILE_STEP = '%s .select-server-profile-step' % MODAL
    OVERRIDE_BUTTON = '%s .override' % MODAL
    PROCEED_BUTTON = '%s .proceed button' % MODAL
    HOST_ADDRESS_TEXT = '%s .pdsh-input input' % MODAL
    SUCCESS_BUTTON = '%s .btn-success' % MODAL

    @property
    def override_button(self):
        return self.driver.find_element_by_css_selector(self.OVERRIDE_BUTTON)

    @property
    def proceed_button(self):
        return self.driver.find_element_by_css_selector(self.PROCEED_BUTTON)

    def wait_for_proceed_enabled(self):
        def wait_for_enabled(add_server_modal):
            try:
                return add_server_modal.proceed_button.is_enabled()
            except StaleElementReferenceException:
                return False

        return WebDriverWait(self, self.medium_wait).until(wait_for_enabled)

    def submit_address(self):
        self.driver.find_element_by_css_selector(self.SUCCESS_BUTTON).click()
        self.wait_for_angular()

    def select_profile(self, profile_text='Managed storage server'):
        self.wait_for_angular()
        profile_select = FancySelect(self.driver, self.SELECT_SERVER_PROFILE_STEP)
        profile_select.select_option(profile_text)

    def submit_profile(self):
        try:
            self.override_button.click()
        except NoSuchElementException:
            pass

        self.proceed_button.click()
        self.wait_for_angular()

    def enter_address(self, host_name):
        self.wait_for_angular()
        enter_text_for_element(self.driver, self.HOST_ADDRESS_TEXT, host_name)

        script = """
            var callback = arguments[arguments.length - 1];

            if (hasScopeUpdated())
              callback(true);
            else
              retry(30);

            function retry (n) {
                if (hasScopeUpdated())
                  callback(true);
                else if (n < 1)
                   callback(false);
                else
                  window.setTimeout(function () { retry(n - 1); }, 1000);
            };

            function hasScopeUpdated () {
              return $('%s').scope().addServer.fields.pdsh === '%s';
            }
        """ % (self.MODAL_BODY, host_name)

        return self.driver.execute_async_script(script)

File: chroma-manager/chroma_ui_new/test/selenium/views/modal.js
'use strict';

var format = require('util').format;
var config = require('../config');
var _ = require('lodash');
var fancySelect = require('./fancy-select');
var isRemoved = require('../util/is-removed');
var msToSec = require('../util/time').msToSec;

var modal = {
  get modal () {
    return $(this.MODAL);
  },
  get modalBackdrop () {
    return $(this.MODAL_BACKDROP);
  },
  modalIsDisplayed: function modalIsDisplayed () {
    return this.modal.isDisplayed();
  },
  /**
   * Wait for a modal to be present.
   * @param {Number} [waitTime]
   */
  waitForModal: function waitForModal (waitTime) {

    waitTime = waitTime || config.wait_time.medium;

    var that = this;

    browser.wait(
      waitOnModal,
      waitTime,
      format('Modal[%j] still present after %d seconds', that.MODAL, msToSec(waitTime))
    );

    /**
     * Wait for the modal.
     * @returns {Object}
     */
    function waitOnModal () {
      return browser.isElementPresent(that.modal);
    }

  },
  /**
   * Wait for the removal of any modal that has it's prototype
   * pointing to this object literal.
   * @param {Number} waitTime
   */
  waitForModalRemove: function waitForModalRemove (waitTime) {

    waitTime = waitTime || config.wait_time.medium;

    var modalIsRemoved = isRemoved(this.modal);

    browser.wait(
      modalIsRemoved,
      waitTime,
      format('Modal[%s] still present after %d seconds', this.MODAL, msToSec(waitTime))
    );
  }
};

var addServerModal = Object.create(modal);
_.extend(addServerModal, {
  get PDSH_EXPRESSION () {

    return config.lustre_servers.reduce(buildPdshExpression, '');

    /**
     * Combine all the nodenames into one expression.
     * @param {String} prev
     * @param {Object} curr
     * @param {Number} idx
     * @returns {String}
     */
    function buildPdshExpression (prev, curr, idx) {
      if (idx === 0)
        return curr.nodename;
      else
        return prev + ',' + curr.nodename;
    }
  },
  MODAL: '.add-server-modal',
  MODAL_BACKDROP: '.add-server-modal-backdrop',
  get MODAL_BODY () {
    return format('%s .modal-body', addServerModal.MODAL);
  },
  get SELECT_SERVER_PROFILE_ () {
    return format('%s .select-server-profile-step', addServerModal.MODAL);
  },
  get OVERRIDE_BUTTON () {
    return format('%s .override', addServerModal.MODAL);
  },
  get PROCEED_BUTTON () {
    return format('%s .proceed button i', addServerModal.MODAL);
  },
  get HOST_ADDRESS_TEXT () {
    return format('%s .pdsh-input input', addServerModal.MODAL);
  },
  get SUCCESS_BUTTON () {
    return format('%s .btn-success', addServerModal.MODAL);
  },
  get overrideButton () {
    return $(addServerModal.OVERRIDE_BUTTON);
  },
  get proceedButton () {
    return $(addServerModal.PROCEED_BUTTON);
  },
  get successButton () {
    return $(addServerModal.SUCCESS_BUTTON);
  },
  /**
   * Wait for the proceed button to be enabled.
   */
  waitForProceedEnabled: function waitForProceedEnabled () {
    function waitOnProceedButton () {
      return browser.isElementPresent(addServerModal.proceedButton)
        .then(function isProceedPresentThen (isProceedPresent) {
          return isProceedPresent;
        });
    }

    browser.wait(waitOnProceedButton, config.wait_time.medium);
  },
  get hostAddressText () {
    return element(by.css(addServerModal.HOST_ADDRESS_TEXT));
  },
  enterAddress: function enterAddress () {

    addServerModal.hostAddressText.sendKeys(addServerModal.PDSH_EXPRESSION);

    addServerModal.waitForAddress();

  },
  waitForAddress: function waitForAddress () {

    function waitOnAddressToBeFullyEntered () {
      return addServerModal.hostAddressText.getAttribute('value')
        .then(function checkInputValueMatchesPdshExpression (text) {
          if (text === addServerModal.PDSH_EXPRESSION)
            return true;
        });
    }

    browser.wait(
      waitOnAddressToBeFullyEntered,
      config.wait_time.medium,
      format('Still waiting for the host address input to be %s after %d seconds',
        addServerModal.PDSH_EXPRESSION, msToSec(config.wait_time.medium))
    );
  },
  submitAddress: function submitAddress () {
    return addServerModal.successButton.click();
  },
  selectProfile: function selectProfile () {
    fancySelect.selectOption();
  },
  submitProfile: function submitProfile () {
    addServerModal.overrideButton.click();
    addServerModal.proceedButton.click();
  }
});

var commandModal = Object.create(modal);
_.extend(commandModal, {
  MODAL: '.command-modal',
  MODAL_BACKDROP: '.command-modal-backdrop',
  get CLOSE_BUTTON () {
    return format('%s .btn-danger', commandModal.MODAL);
  },
  get closeButton () {
    return $(commandModal.CLOSE_BUTTON);
  },
  waitForCloseButtonToBeClickable: function waitForCloseButtonToBeClickable () {
    function waitOnCloseButton () {
      return browser.isElementPresent(commandModal.closeButton);
    }

    browser.wait(waitOnCloseButton, config.wait_time.medium);
  }
});

var confirmActionModal = Object.create(modal);
_.extend(confirmActionModal, {
  MODAL: '.confirm-action-modal',
  MODAL_BACKDROP: '.confirm-action-modal-backdrop',
  get CONFIRM_BUTTON () {
    return format('%s .btn-success i:not(dropdown-toggle)', confirmActionModal.MODAL);
  },
  get confirmButton () {
    return $(confirmActionModal.CONFIRM_BUTTON);
  }
});

var serverDetailModal = Object.create(modal);
_.extend(serverDetailModal, {
  MODAL: '.server-detail-modal'
});

module.exports = {
  commandModal: commandModal,
  addServerModal: addServerModal,
  confirmActionModal: confirmActionModal
};
